### PR TITLE
FIX-#6367: Enable support for 'groupby.size()' in reshuffling groupby

### LINKED ISSUE #6367: BUG: Reshuffling groupby doesn't work for `groupby.size()` method
### Modin version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest released version of Modin.

- [X] I have confirmed this bug exists on the main branch of Modin. (In order to do this you can follow [this guide](https://modin.readthedocs.io/en/stable/getting_started/installation.html#installing-from-the-github-master-branch).)

### Reproducible Example

```python
import modin.pandas as pd
import modin.config as cfg

df = pd.DataFrame({"a": [1, 2, 2, 4], "b": [3, 4, 5, 6]})

cfg.ExperimentalGroupbyImpl.put(True)
df.groupby("a").size()
# UserWarning: Can't use experimental reshuffling groupby implementation because of:
# Reshuffling groupby is only supported when grouping on a column(s) of the same frame.
# https://github.com/modin-project/modin/issues/5926
# Falling back to a TreeReduce implementation.
```

### Issue Description

Although we're grouping on the column of the `df` in the example above we still get a warning telling us that the reshuffling is not supported for this case.

This happens because we overwrite the `drop` parameter to `False` at the high-level `groupby.size()` implementation, telling us that the `by` columns should now be considered as external ones, not belonging to the original frame:
https://github.com/modin-project/modin/blob/6a5416c7cbe230bcd3fb359165f1eee535d35658/modin/pandas/groupby.py#L969

### Expected Behavior

`.groupby.size()` to work via reshuffling

### Error Logs

N/A

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

This PR removes `drop=False` parameter overwriting for the groupby object used in `groupby.size()`. The overwriting was originally introduced by this commit bb8b23dae531514a4271eb4395e2f52fdd28c53e more than 3 years ago in an attempt to fix `groupby(as_index=False, ...).size()`. Nowadays, this case works perfectly fine even without this `drop=False` overwriting.

The problem that this overwriting had brought, is that it confused implementations on whether the 'by' columns actually came from the grouping frame or not (`drop=True` means that the grouping column came from the same frame and `drop=False` means that the 'by' columns are foreign).

The experimental groupby implementation only allows for the 'by' columns to originate from the same frame, meaning that this `drop=False` overwriting never allowed us to run `groupby.size()` using the new implementation. This PR fixes that problem.
https://github.com/modin-project/modin/blob/f24ba6f562759c5807f7f6af1db96899d02a00c8/modin/core/storage_formats/pandas/query_compiler.py#L3421-L3425

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6367 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/groupby.py [hunk 1]
```diff
@@ -962,16 +962,8 @@ def size(self):
                 idx_name=self._idx_name,
                 **self._kwargs,
             ).size()
-        work_object = type(self)(
-            self._df,
-            self._by,
-            self._axis,
-            drop=False,
```
Comment (dchigarev) [inline modin/pandas/groupby.py:969]: explicitly setting `drop=False` seems useless for the result of `groupby.size()`

**Review by anmyachev — DISMISSED** [review]
LGTM

File: modin/test/storage_formats/pandas/test_internals.py [hunk 2]
```diff
@@ -1089,3 +1090,22 @@ def test_setitem_bool_preserve_dtypes():
...
+    # https://github.com/modin-project/modin/issues/6367
+    df = pd.DataFrame({"a": [1, 1, 2, 2], "b": [3, 4, 5, 6]})
+    modin_frame = df._query_compiler._modin_frame
+
+    with mock.patch.object(
+        modin_frame,
+        "_apply_func_to_range_partitioning",
+        wraps=modin_frame._apply_func_to_range_partitioning,
+    ) as shuffling_method:
+        repr(df.groupby("a").size())
```
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/test/storage_formats/pandas/test_internals.py]: changed

**Review by anmyachev — DISMISSED** [review]

File: modin/test/storage_formats/pandas/test_internals.py [hunk 3]
```diff
@@ -1089,3 +1091,22 @@ def test_setitem_bool_preserve_dtypes():
...
+    modin_frame = df._query_compiler._modin_frame
+
+    with mock.patch.object(
+        modin_frame,
+        "_apply_func_to_range_partitioning",
+        wraps=modin_frame._apply_func_to_range_partitioning,
+    ) as shuffling_method:
+        try_cast_to_pandas(df.groupby("a").size())
+
+    shuffling_method.assert_called()
```
Comment (vnlitvinov) [inline modin/test/storage_formats/pandas/test_internals.py:1112]: what assert are we expecting there?

Comment (dchigarev) [reply modin/test/storage_formats/pandas/test_internals.py:1112]: The test verifies that setting the `ExperimentalGroupbyImpl` to True will guarantee us to engage the reshuffling groupby implementation. 

[The patched method](https://github.com/modin-project/modin/blob/947e06bb07ae93633e2b0a68ea9d320f689d5274/modin/core/dataframe/pandas/dataframe/dataframe.py#L2289-L2310) is the one that actually builds range-partitioning and executes functions over it, so if the code made it through to this method, we can assume that the reshuffling implementation was used, rather [than the default MapReduce one.](https://github.com/modin-project/modin/blob/947e06bb07ae93633e2b0a68ea9d320f689d5274/modin/core/storage_formats/pandas/query_compiler.py#L3267-L3271)

**Review by anmyachev — APPROVED** [review]

**Comment by anmyachev** [comment]
@vnlitvinov do you have any other comments?

**Pull request merged**